### PR TITLE
[Feat] Single Batch Overlap (SBO): Overlaping of Down GEMM with Combine Send

### PR DESCRIPTION
**Pull request by Sulfur6**
# 1. Motivation
The optimization effect of Two-Batch Overlap (TBO) is suboptimal for the Decode phase on low-compute-power cards (i.e., H20). This is due to two main factors: First, on the Hopper architecture, the WGMMA block_m is 64. Consequently, when TBO is enabled with a small Decode batch size, the MLP GEMM suffers from redundant computations. A positive throughput gain is only observed at larger batch sizes (e.g., 64, 128). Second, at these larger batch sizes, low-compute-power cards like the H20 fail to meet the SLA guarantees for TPOT/ITL.

Therefore, it is necessary to find a solution that can improve Decode throughput even with small batch sizes. Single Batch Overlap (SBO) presents itself as a viable solution.

We implement SBO for DeepSeek v3/R1 by modifying DeepEP and DeepGEMM, including the overlap of Shared Expert and Dispatch Recv, as well as the overlap of Down GEMM with Combine Send.

The overlap of Down GEMM with Combine Send is implemented by modifying SGlang, DeepEP and DeepGEMM, with the detailed implementation available in the PRs below:
* DeepEP: https://github.com/deepseek-ai/DeepEP/pull/390.
* DeepGEMM: this PR.

We also conducted integration and evaluation in SGLang: https://github.com/sgl-project/sglang/pull/9660.

# 2. Overlap Design
SBO implements two overlap for the MoE layers of DeepSeek-V3/R1. One is to overlap the Shared Expert computation with the Dispatch Recv communication, and the other is to overlap the Down GEMM computation with the Combine Send communication.
image
The interaction between Down GEMM and Combine Send is structured as a producer-consumer model synchronized by signals. For each local expert, a signal unit is allocated for every block_m tokens. The Down GEMM computes the results for these block_m tokens and atomically increments the signaling unit after completing a portion of the work. The Combine Send polls this signaling unit. Once the value reaches a threshold, it sends the corresponding block_m tokens.
image

# 3. Modifications
* Add the `m_grouped_fp8_gemm_nt_signal` Python interface to support overlapping Down GEMM with Combine Send.
* Add the `SM90FP8SignalGemm1D2DRuntime` and `sm90_m_grouped_fp8_gemm_signal_1d2d` to support Signal Down GEMM in SM90.
* The `sm90_fp8_signal_gemm_1d2d_impl` kernel uses `atomicAdd` to write signal after the corresponding block_m tokens are computed.

# 4. Evaluation
We integrated the modified DeepEP and DeepGEMM into SGLang for performance evaluation.

## 4.1. Experiment Setup
* 5 nodes, with 8 × H20 GPUs per node. Each prefill node uses TP8, and the other 2 decode nodes use DP_Attn 16 + EP 16.
* Input length 4096, output length 1536.

## 4.2. Performance Evaluation
The following is the evaluation results based on DeepGEMM v1. SBO has adapted SGLang and DeepGEMM v2, and is evaluating. The results will be updated soon.

* bs 32, origin
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    4.8
Max request concurrency:                 512
Successful requests:                     10240
Benchmark duration (s):                  2343.93
Total input tokens:                      41943040
Total generated tokens:                  15728640
Total generated tokens (retokenized):    15673809
Request throughput (req/s):              4.37
Input token throughput (tok/s):          17894.29
Output token throughput (tok/s):         6710.36
Total token throughput (tok/s):          24604.65
Concurrency:                             489.92
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   112143.54
Median E2E Latency (ms):                 113007.21
---------------Time to First Token----------------
Mean TTFT (ms):                          1052.21
Median TTFT (ms):                        939.13
P99 TTFT (ms):                           2528.55
---------------Inter-Token Latency----------------
Mean ITL (ms):                           72.37
Median ITL (ms):                         71.14
P95 ITL (ms):                            89.53
P99 ITL (ms):                            150.60
Max ITL (ms):                            1022.47
==================================================
```
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    5.0
Max request concurrency:                 512
Successful requests:                     10240
Benchmark duration (s):                  2346.74
Total input tokens:                      41943040
Total generated tokens:                  15728640
Total generated tokens (retokenized):    15673126
Request throughput (req/s):              4.36
Input token throughput (tok/s):          17872.88
Output token throughput (tok/s):         6702.33
Total token throughput (tok/s):          24575.22
Concurrency:                             490.96
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   112515.52
Median E2E Latency (ms):                 113174.24
---------------Time to First Token----------------
Mean TTFT (ms):                          1100.37
Median TTFT (ms):                        950.60
P99 TTFT (ms):                           2931.98
---------------Inter-Token Latency----------------
Mean ITL (ms):                           72.58
Median ITL (ms):                         71.20
P95 ITL (ms):                            89.90
P99 ITL (ms):                            150.52
Max ITL (ms):                            912.97
==================================================
```

* bs 32, sbo
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    4.8
Max request concurrency:                 512
Successful requests:                     10240
Benchmark duration (s):                  2235.23
Total input tokens:                      41943040
Total generated tokens:                  15728640
Total generated tokens (retokenized):    15673167
Request throughput (req/s):              4.58
Input token throughput (tok/s):          18764.56
Output token throughput (tok/s):         7036.71
Total token throughput (tok/s):          25801.26
Concurrency:                             485.50
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   105977.11
Median E2E Latency (ms):                 106863.39
---------------Time to First Token----------------
Mean TTFT (ms):                          1275.84
Median TTFT (ms):                        1151.83
P99 TTFT (ms):                           3623.72
---------------Inter-Token Latency----------------
Mean ITL (ms):                           68.21
Median ITL (ms):                         67.05
P95 ITL (ms):                            84.90
P99 ITL (ms):                            141.44
Max ITL (ms):                            829.91
==================================================
```
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    5.0
Max request concurrency:                 512
Successful requests:                     10240
Benchmark duration (s):                  2226.19
Total input tokens:                      41943040
Total generated tokens:                  15728640
Total generated tokens (retokenized):    15673172
Request throughput (req/s):              4.60
Input token throughput (tok/s):          18840.70
Output token throughput (tok/s):         7065.26
Total token throughput (tok/s):          25905.96
Concurrency:                             489.36
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   106386.88
Median E2E Latency (ms):                 106958.25
---------------Time to First Token----------------
Mean TTFT (ms):                          1205.13
Median TTFT (ms):                        1109.33
P99 TTFT (ms):                           3184.52
---------------Inter-Token Latency----------------
Mean ITL (ms):                           68.52
Median ITL (ms):                         67.14
P95 ITL (ms):                            85.35
P99 ITL (ms):                            142.63
Max ITL (ms):                            1225.78
==================================================
```

## 4.3. Accuracy Tests
```
#python -m benchmark.gsm8k.bench_sglang --port 8000 --num-questions 1000
100%|█████████████████████████████████████████████████████████████| 1000/1000 [01:22<00:00, 12.17it/s]
Accuracy: 0.951
Invalid: 0.000
Latency: 82.471 s
Output throughput: 1165.550 token/s
```

## 4.4. Repro Script
Please refer to https://github.com/sgl-project/sglang/pull/9660.

### CODE REVIEW

File: deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh [hunk 1]
```diff
@@ -0,0 +1,457 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (Sulfur6) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh:1]: OK, I will make the changes as soon as possible based on your suggestions. BTW, I'd like to know whether you recommend merging the interface/impl/runtime and kernel into the existing content, or just merging the kernel into the existing kernel?

[the remark this answers is not in the document]
Comment (fzyzcjy) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh:1]: I personally recommend that (same for deepep & deepgemm & sglang pr), when code are largely similar and only differ by a small portion, it would be great to merge them

as for this PR, my imagination is like, when looking at code diff, it is very tiny and only contain e.g. several dozen lines of code like "passing around the signal tensor" and "atomic change the signal", instead of ~1k line of code diff

[the remark this answers is not in the document]
Comment (Sulfur6) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh:1]: OK, I will make the changes as soon as possible based on your suggestions.

[the remark this answers is not in the document]
Comment (fzyzcjy) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh:1]: looking forward to it

[the remark this answers is not in the document]
Comment (fzyzcjy) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh:1]: feel free to ping me when there is updates, then I will start working on it

[the remark this answers is not in the document]
Comment (Sulfur6) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh]: I have refactored the code according to your suggestions, reducing code duplication as much as possible. I will further adapt the modified DeepGEMM to sglang, reduce code duplication in sglang, and refactor the code style.

[the remark this answers is not in the document]
Comment (fzyzcjy) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh]: wow great

[the remark this answers is not in the document]
Comment (fzyzcjy) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh]: for sglang part, maybe we should somehow unify my branch https://github.com/sgl-project/sglang/pull/9870 and your branch. (I created that branch b/c needed it immediately and have sth different from yours, e.g. I need to try overlap shared w/ combine)

my personal suggestion is, maybe your branch can wait a bit and just leave it as-is (as long as it can run correct accuracy and speed), and later I will refactor my branch to provide places where your changes can be easily added in, and then after my branch is merged your branch can be refactored with hopefully minor changes

File: tests/test_signal_gemm.py
Comment (fzyzcjy) [inline tests/test_signal_gemm.py:1]: tiny nit: seems we have test_m_grouped_gemm_signal in test_fp8.py, maybe this file forget to be removed

**Review by fzyzcjy — COMMENTED** [review]
(I will have a more detailed check a bit later)

File: deep_gemm/include/deep_gemm/impls/sm90_fp8_gemm_1d2d.cuh [hunk 2]
```diff
@@ -428,6 +430,21 @@ sm90_fp8_gemm_1d2d_impl(float* sfb, int* grouped_layout,
...
+                    cute::tma_store_wait<0>();
+                }
+
+                cg::coalesced_group group = cg::coalesced_threads();
+                group.sync();
+
+                __threadfence();
+
+                if (threadIdx.x == 0) {
+                    atomicAdd(signal + scheduler.current_group_idx * ceil_div(shape_m, BLOCK_M) + m_block_idx, 1);
```
Comment (fzyzcjy) [inline deep_gemm/include/deep_gemm/impls/sm90_fp8_gemm_1d2d.cuh:445]: btw I am still a bit worried about this atomicAdd... (the code location issue in https://github.com/sgl-project/sglang/pull/9660#issuecomment-3230244330 is already solved and no problem)

File: deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sulfur6) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_signal_gemm_1d2d.cuh]: Thank you very much, I will keep my sglang branch https://github.com/sgl-project/sglang/pull/9660 all working fine, and then wait to unify with your branch.

File: tests/test_signal_gemm.py
Comment (Sulfur6) [reply tests/test_signal_gemm.py:1]: I will deal with these issues as soon as possible.

File: deep_gemm/include/deep_gemm/impls/sm90_fp8_gemm_1d2d.cuh (hunk 2 above)
Comment (fzyzcjy) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_gemm_1d2d.cuh:441]: My naive understanding is the following (please correct me if I am wrong!):

> the thread initiating the TMA operation and the thread executing the write signal are not necessarily the same

`cutlass::arch::NamedBarrier(kNumMathThreads).sync();` ensures all math threads sync at this point, thus we know all tma_store_wait calls have finished at this point, thus when we do a `atom.add.release.gpu.global.s32` on thread 0, we already ensure tma_store_wait from thread 0,1,2,3,... all done.

File: deep_gemm/include/deep_gemm/impls/sm90_fp8_gemm_1d2d.cuh [hunk 3]
```diff
@@ -428,6 +428,19 @@ sm90_fp8_gemm_1d2d_impl(float* sfb, int* grouped_layout,
                 cute::tma_store_arrive();
             }
             __syncwarp();
+
+            if constexpr (kEnableOverlap) {
+                if (threadIdx.x < BLOCK_N / TMA_D_BLOCK_N) {
+                    cute::tma_store_wait<0>();
```
Comment (fzyzcjy) [inline deep_gemm/include/deep_gemm/impls/sm90_fp8_gemm_1d2d.cuh:434]: one more naive worry: `tma_store_wait` seems to correspond to `cp.async.bulk.wait_group.read` (src: https://github.com/NVIDIA/cutlass/blob/76c96b0be35cb263debe3e3d8418b80911a544ab/include/cute/arch/copy_sm90_tma.hpp#L1251), but it seems that we need `cp.async.bulk.wait_group` (no ".read"), otherwise the semantics (https://docs.nvidia.com/cuda/parallel-thread-execution/#data-movement-and-conversion-instructions-cp-async-bulk-wait-group) is that, "the tma store has done reading from source, but the `Writes being made visible to the executing thread` may not have been done".

Comment (fzyzcjy) [reply deep_gemm/include/deep_gemm/impls/sm90_fp8_gemm_1d2d.cuh:434]: looking forward to the fix

File: deep_gemm/testing/numeric.py [hunk 4]
```diff
@@ -1,6 +1,20 @@
 import torch
 from typing import Iterable
 
+def check_signal(num_local_expert, max_m, block_m, threshold, signal, masked_m):
+    ceil_div = lambda a, b: (a + b - 1) // b
+
+    expert_len = max_m // block_m
```
Comment (qq332982511) [inline deep_gemm/testing/numeric.py]: `expert_len = max_m // block_m` -> `expert_len = (max_m + block_m - 1) // block_m` ?
